salt/modules/yumpkg.py

```diff
@@ -3217,3 +3217,76 @@ def list_installed_patches():
         salt '*' pkg.list_installed_patches
     '''
     return _get_patches(installed_only=True)
+
+
+@salt.utils.decorators.path.which('yum-complete-transaction')
+def complete_transaction(cleanup_only=False, recursive=False, max_attempts=3):
+    '''
+    .. versionadded:: Fluorine
+
+    Execute ``yum-complete-transaction``, which is provided by the ``yum-utils`` package.
+
+    cleanup_only
+        Specify if the ``--cleanup-only`` option should be supplied.
+
+    recursive
+        Specify if ``yum-complete-transaction`` should be called recursively
+        (it only completes one transaction at a time).
+
+    max_attempts
+        If ``recursive`` is ``True``, the maximum times ``yum-complete-transaction`` should be called.
+
+    .. note::
+
+        Recursive calls will stop once ``No unfinished transactions left.`` is in the returned output.
+
+    .. note::
+
+        ``yum-utils`` will already be installed on the minion if the package
+        was installed from the Fedora / EPEL repositories.
+
+    CLI example:
+
+    .. code-block:: bash
+
+        salt '*' pkg.complete_transaction
+        salt '*' pkg.complete_transaction cleanup_only=True
+        salt '*' pkg.complete_transaction recursive=True max_attempts=5
+    '''
+
+    return _complete_transaction(cleanup_only, recursive, max_attempts, 1, [])
+
+
+def _complete_transaction(cleanup_only, recursive, max_attempts, run_count, cmd_ret_list):
+    '''
+    .. versionadded:: Fluorine
+
+    Called from ``complete_transaction`` to protect the arguments
+    used for tail recursion, ``run_count`` and ``cmd_ret_list``.
+    '''
+
+    cmd = ['yum-complete-transaction']
+    if cleanup_only:
+        cmd.append('--cleanup-only')
+
+    cmd_ret_list.append(__salt__['cmd.run_all'](
+        cmd,
+        output_loglevel='trace',
+        python_shell=False
+    ))
+
+    if cmd_ret_list[-1]['retcode'] == 0 \
+        and recursive \
+            and 'No unfinished transactions left.' not in cmd_ret_list[-1]['stdout']:
+        if run_count >= max_attempts:
+            cmd_ret_list[-1]['retcode'] = 1
+            if cmd_ret_list[-1]['stderr']:
+                # this shouldn't happen, since retcode was 0 and we had no unfinished transactions
+                cmd_ret_list[-1]['stderr'] += '\n'
+            cmd_ret_list[-1]['stderr'] += 'Command succeeded but attempt {0}/{1} exceeded max.' \
+                                          ' Unfinished transactions remain.'.format(run_count, max_attempts)
```

Comment:
Usually this is `comment` or `changes` section for `STDOUT` and `errors` for `STDERR`. And the [line continuation](http://docs.python-guide.org/en/latest/writing/style/#line-continuations) applies here too.

Reply:
Yeah, `stderr` is usually reserved for the output of a command. What we should be doing here actually is raising a `CommandExecutionError`. This exception class provides for an optional `info` parameter, which will then have its results rendered in the return:

```python
>>> # assuming max_attempts == 1
...
>>> from salt.exceptions import CommandExecutionError
>>> cmd_ret_list = [{'pid': 12345, 'retcode': 0, 'stdout': 'Applied transaction', 'stderr': ''}, {'pid': 23456, 'retcode': 0, 'stdout': 'Applied transaction', 'stderr': ''}]
>>> try:
...     raise CommandExecutionError('Command succeeded but attempts exceeded max', info={'results': results})
... except CommandExecutionError as exc:
...     pass
...
>>> print(exc)
Command succeeded but attempts exceeded max. Additional info follows:

results:
    |_
      ----------
      pid:
          12345
      retcode:
          0
      stderr:
      stdout:
          Applied transaction
    |_
      ----------
      pid:
          23456
      retcode:
          0
      stderr:
      stdout:
          Applied transaction
>>> exc.info
{'results': [{'pid': 12345, 'retcode': 0, 'stderr': '', 'stdout': 'Applied transaction'}, {'pid': 23456, 'retcode': 0, 'stderr': '', 'stdout': 'Applied transaction'}]}
```
So the string that I printed there is what you'd see in the return from that minion, but you'd also have the ability to catch that exception and access the results. That's not something that would be necessary in this case, but it does leave that option open to any code which invokes this function.